### PR TITLE
feat(mutators): Add Math methods mutator

### PR DESCRIPTION
**Pull request by yakivyusin**
Mutator from #1194 .
In addition to the proposed swaps from the issue, the implementation covers hyperbolic, inverse hyperbolic, inverse trigonometric functions, and some functions from newer .NET releases.

### CODE REVIEW

**Comment by rouke-broersma** [comment]
Cool! Thank you :)

**Comment by rouke-broersma** [comment]
@richardwerkman since you seem to have more affinity with math, could you review this new mutator?

File: docs/mutations.md [hunk 1]
```diff
@@ -159,3 +159,56 @@ Do you have a suggestion for a (new) mutator? Feel free to create an [issue](htt
 
 ## Regular Expressions (_regex_)
 For the full list of all available regex mutations, see the [regex mutator docs](./regex-mutations.md).
+
+## Math Methods (_math_)
+|      Original           |         Mutated         |
+| ----------------------- | ----------------------- |
+| Acos()                  | Acosh()                 |
+| Acos()                  | Asin()                  |
+| Acos()                  | Atan()                  |
```
Comment (psfinaki) [inline docs/mutations.md:168]: Is there a specific reason to have multiple mutated variants for the same thing? 

I vaguely remember from my trigonometry lessons at school 15 years ago that there are some cases when some of the pairs will have the same values (like errrrhhhhh sin and cos for 45 degrees?) but I think philosophically Stryker in the vast majority of cases puts up with that (@rouke-broersma correct me if I am wrong).

Like, Stryker changes + to - and other way round, but it doesn't also change + to * or /, even though `x + 0` and `x - 0` will give the same result. AFAIU the framework kind of tries to find a balance between the amount of value and amount of generated mutants (because the amount eventually affects the performance).

TL;DR - I'd stick with one mutant for each expression.

File: src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            while (node is ConditionalAccessExpressionSyntax conditional)
+            {
+                foreach (var subMutants in FindMutableMethodCalls(conditional.Expression, original))
+                {
+                    yield return subMutants;
+                }
+                node = conditional.WhenNotNull;
+            }
+
+            for (; ; )
```
Comment (psfinaki) [inline src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:71]: First, that's an obscure way to create an infinite loop from the times of C, I'd prefer any clearer form, e.g. `while (true)`.

Second, I haven't created mutators myself but I don't think I've seen infinite loops in the mutators' code (@dupdob correct me if I am wrong) and so I'm just generally paranoid about them. Can you somehow verbally describe this need here? Or maybe extract a method and give it a very expressive name :)

**Review by psfinaki — COMMENTED** [review]
Got a bit of time, looked through the code :) thanks - it was a nice journey into System.Math, which I have never had an opportunity for in my career heh.

File: docs/mutations.md (hunk 1 above)
Comment (yakivyusin) [reply docs/mutations.md:168]: Hello.

The case with `sin(x) = cos(x)` was discussed in the initial issue - they are indeed equal for 45 and 225 degrees. From my experience with mutation testing, if someone gets such survived mutant then this is a good point to improve tests - just add a new test case with an another angle.

The same for other swaps, they will test well other "magic" angles that can be used in incomplete tests - e.g. `sin(x) = sinh(x)` for 0 and `sin(x) = tan(x)` for `pi*n`.

And swap trigonometric <-> hyperbolic function looks like a real possible error with the function selection in IntelliSense :D

Comment (psfinaki) [reply docs/mutations.md:168]: Alright. I mean, I haven't worked in codebases where this stuff was heavily used, just wanted to share the concern that it can generate many mutants. But yeah, that's an advanced mutator anyway and will probably not be applicable for most repos.

File: src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs (hunk 2 above)
Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:71]: Hmm, alright, good to know! But my feedback (and I am just a random dev) is the same - the code could be clearer here, we don't need to be uniform in crypticity ;)

Comment (yakivyusin) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:71]: Yes, sure, I will think about how to refactor this place. Then, maybe, the good idea will be to refactor the same place in `LinqMutator` in the same way.

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:71]: I understand why you picked `LinqMutator` as a template, but it is needlessly complex for Math methods: as those are static methods, you do not need to worry about conditional accessor or chained invocation. You also don't need to worry about MemberBinding expression.
So you need to look for invocation looking like `Math.[Sin|Cos|Tan|...]`, which are `InvocationExpression` using  a `SimpleMemberAcessExpression` as its Expression. This one has an `IdentifierName` (token 'Math') as an expression and another `IndentifierName`(identifier 'Sin', 'Cos', ...).

Therefore you can drop the initial `while` loop and transform the switch case to a simpler `if`.

You can use sharplab.io to have an idea of how different syntax constructs are represented with Roslyn.
Here is a [simple example](https://sharplab.io/#v2:C4LghgzsA0AmIGoA+ABATARgLACgUGYACdQgYUIG9dCbiiUAWQgWQAoBKS62n5sYABYA6AMoBLAHasMABnYBubjQC+uZUA==).
I think you will find this useful.

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/MathMutatorTest.cs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Shouldly;
+using Stryker.Core.Mutators;
+using Xunit;
+
+namespace Stryker.Core.UnitTest.Mutators
+{
+    public class MathMutatorTest : TestBase
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Mutators/MathMutatorTest.cs:13]: I would like to see a test to see if these methods are *not* mutated when the class has a different name. I can imagine we will see a class with a method Ceiling and we could mutate that as well (which will fail)

File: src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            [MathExpression.ReciprocalEstimate] = new[] { MathExpression.ReciprocalSqrtEstimate },
+            [MathExpression.ReciprocalSqrtEstimate] = new[] { MathExpression.ReciprocalEstimate, MathExpression.Sqrt },
+            [MathExpression.Sin] = new[] { MathExpression.Sinh, MathExpression.Cos, MathExpression.Tan },
+            [MathExpression.Sinh] = new[] { MathExpression.Sin, MathExpression.Cosh, MathExpression.Tanh },
+            [MathExpression.Tan] = new[] { MathExpression.Tanh, MathExpression.Cos, MathExpression.Sin },
+            [MathExpression.Tanh] = new[] { MathExpression.Tan, MathExpression.Cosh, MathExpression.Sinh }
+        };
+
+        /// <summary> Apply mutations to an <see cref="InvocationExpressionSyntax"/> </summary>
+        public override IEnumerable<Mutation> ApplyMutations(ExpressionSyntax node)
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:46]: Can't we target a more specific Syntax type? This mutator will now be called for almost every type of node...

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/MathMutatorTest.cs (hunk 3 above)
Comment (yakivyusin) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/MathMutatorTest.cs:13]: What about `using static System.Math` or `using AnyName = System.Math`?

File: src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs (hunk 2 above)
Comment (yakivyusin) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:71]: Thank you. I will change the implementation to use this way.

File: src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs (hunk 4 above)
Comment (yakivyusin) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:46]: Thank you, I will limit method to `InvocationExpressionSyntax` during refactoring.

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/MathMutatorTest.cs (hunk 3 above)
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/MathMutatorTest.cs:13]: I guess you should handle `using static System.Math` which implies supporting the short call form `Cos(x)` : easy to do and with limited compilation errors risk (those should only happens WITHIN a class providing methods with mathematical names).

I think the alias form is so rare there is no need to support it. Remember: mutation testing is about stress testing your tests, this is not some form of extensive testing.

It should be easy to properly support aliasing when the decision is made to generate semantic information alongside the syntax tree (if ever).

Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/MathMutatorTest.cs:13]: > I think the alias form is so rare there is no need to support it.

Murphy's law says that whatever can happen will happen :D

File: src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        {
+            MemberAccessExpressionSyntax memberAccess => ApplyMutationsToMemberCall(node, memberAccess),
+            IdentifierNameSyntax methodName => ApplyMutationsToDirectCall(node, methodName),
+            _ => Enumerable.Empty<Mutation>()
+        };
+
+        private static IEnumerable<Mutation> ApplyMutationsToMemberCall(InvocationExpressionSyntax node, MemberAccessExpressionSyntax memberAccessExpressionSyntax)
+        {
+            if (memberAccessExpressionSyntax.Expression is not IdentifierNameSyntax memberName ||
+                !memberName.Identifier.ValueText.StartsWith("Math"))
```
Comment (psfinaki) [inline src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:57]: Maybe at least it's possible to verify the "System" namespace here for it?

**Review by psfinaki — APPROVED** [review]

File: src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs (hunk 2 above)
Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs]: Anyway, code is much clearer now, good job!

File: src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs (hunk 5 above)
Comment (yakivyusin) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:57]: From my point of view, it looks like we can't check the fully qualified name without the semantic model. But maybe I'm wrong.

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:57]: yes and no; you can check the given name, so you can verify that the name is either `Math.Con` or `Con` (no namespace provided) not 'Something.Con` nor `Something.Math.Con`. Less precise than using the semantic model, but good enough for most cases

Comment (yakivyusin) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:57]: @dupdob But the existing code already works like that, doesn't it? The code exactly checks what you described.

Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/Mutators/MathMutator.cs:57]: Alrighty then, you can resolve this.

**Comment by yakivyusin** [comment]
@richardwerkman @dupdob Hello, could you please review the new implementation to resolve the old comments & accept the PR?

**Review by dupdob — APPROVED** [review]
Comments have been made and remarks integrated.
Thanks for this good PR PR.
